rust/src/go/nmstate/nmstate.go

```diff
@@ -0,0 +1,199 @@
+// +build cgo
+package nmstate
+
+// #cgo CFLAGS: -g -Wall
+// #cgo LDFLAGS: -lnmstate
+// #include <nmstate.h>
+// #include <stdlib.h>
+import "C"
+import (
+	"fmt"
+	"io"
+	"time"
+)
+
+type Nmstate struct {
+	Timeout	    uint
+	LogsWritter io.Writer
+	Flags	    byte
+	State	    string
+	Checkpoint  string
+}
+
+const (
+	none byte = 0 << iota
+	kernel_only
+	no_verify
+	include_status_data
+	include_secrets
+	no_commit
+)
+
+func New() *Nmstate {
+	return &Nmstate{}
+}
+
+func WithTimeout(timeout time.Duration) func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Timeout = uint(timeout.Seconds())
+	}
+}
+
+func WithLogsWritter(log_writter io.Writer) func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.LogsWritter = log_writter
+	}
+}
+
+func WithKernelOnly() func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Flags = n.Flags | kernel_only
+	}
+}
+
+func WithNoVerify() func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Flags = n.Flags | no_verify
+	}
+}
+
+func WithIncludeStatusData() func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Flags = n.Flags | include_status_data
+	}
+}
+
+func WithIncludeSecrets() func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Flags = n.Flags | include_secrets
+	}
+}
+
+func WithNoCommit() func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Flags = n.Flags | no_commit
+	}
+}
+
+func WithState(state string) func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.State = state
+	}
+}
+
+func WithCheckpoint(checkpoint string) func(*Nmstate) {
+	return func(n *Nmstate) {
+		n.Checkpoint = checkpoint
+	}
+}
+
+// Retrieve the network state in json format. This function returns the current
+// network state or an error.
+func (n *Nmstate) RetrieveNetState() (string, error) {
+	var (
+		state    *C.char
+		log      *C.char
+		err_kind *C.char
+		err_msg  *C.char
+	)
+	rc := C.nmstate_net_state_retrieve(C.uint(n.Flags), &state, &log, &err_kind, &err_msg)
+
+	defer func() {
+		C.nmstate_net_state_free(state)
+		C.nmstate_err_msg_free(err_msg)
+		C.nmstate_err_kind_free(err_kind)
+		C.nmstate_log_free(log)
+	}()
+	_, err := io.WriteString(n.LogsWritter, C.GoString(log))
+	if err != nil {
+		return "", fmt.Errorf("failed writting logs: %v", err)
+	}
+	if rc != 0 {
+		return "", fmt.Errorf("failed retrieving nmstate net state with rc: %d", rc)
```

Comment:
We should add `err_msg`and `err_kind` here right ?

Reply:
Done, thank you!